CUMULUS-1911:Add Reconciliation Reports to Elasticsearch

jennyhliu opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-1911: Add Reconciliation Reports to Elasticsearch](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1911)

## Changes

  - Added ReconciliationReports table
  - Updated CreateReconciliationReport lambda to save Reconciliation Report records to database
  - Updated dbIndexer and IndexFromDatabase lambdas to index Reconciliation Report records to Elasticsearch
  - Added migration_5 to migrate existing s3 reconciliation report records to database and Elasticsearch
  - Updated `@cumulus/api` package, `tf-modules/archive` and `tf-modules/data-persistence` Terraform modules
- Added migration instruction

## PR Checklist

- [x ] Update CHANGELOG
- [x ] Unit tests
- [x ] Adhoc testing
- [x ] Integration tests

yjpa7145 commented:
Now that we moved to Terraform, I don't _believe_ we brought migrations with that. I think that was a `kes` thing. The lambda function looks like it was migrated, but I don't believe it is being invoked anywhere.

jennyhliu commented:
> Now that we moved to Terraform, I don't _believe_ we brought migrations with that. I think that was a `kes` thing. The lambda function looks like it was migrated, but I don't believe it is being invoked anywhere.

Yeah, I just removed the migrations which are no longer used from the lambda.  We can still use the lambda for migration purpose.

yjpa7145 commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -55,7 +55,13 @@ function getReportsKeys(systemBucket, stackName) {
     .then((response) => response.Contents.map((o) => o.Key));
 }
 
+// TODO we will call API to delete the reports when API is updated
```
Is there a separate ticket (or PR) to update the API?

jennyhliu replied:
Yes. CUMULUS-1912

laurenfrederick commented on CHANGELOG.md:
```diff
@@ -14,11 +14,22 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
+- Migrate existing s3 reconciliation report records to database (CUMULUS-1911)
+  After update your Cumulus stack, run the command:
+  ./node_modules/.bin/cumulus-api migrate --stack <your-cloudformation-stack-name> --migrationVersion migration5
```
Can you also make a note please to deploy the DP deployment?

jennyhliu replied:
Where do I put a note?

laurenfrederick replied:
Just in the changelog migration notes. I'll add a suggestion.

laurenfrederick commented on packages/api/tests/lambdas/create-reconciliation-report.js:
```diff
@@ -101,27 +102,25 @@ function storeCollectionsToDynamoDb(tableName, collections) {
-
-  if (report.status === 'RUNNING') {
+async function fetchCompletedReport(t) {
+  const reportRecord = await new models.ReconciliationReport().scan();
```
We shouldn't need to do a scan here, we should be able to do a get by report name

jennyhliu replied:
I don't have a report name.

jennyhliu replied:
Ok.  Updated the lambda to return the report record

laurenfrederick commented on packages/api/tests/lambdas/create-reconciliation-report.js:
```diff
@@ -101,27 +102,25 @@ function storeCollectionsToDynamoDb(tableName, collections) {
+async function fetchCompletedReport(t) {
+  const reportRecord = await new models.ReconciliationReport().scan();
+  t.is(reportRecord.Items.length, 1);
+  if (reportRecord.Items[0].status === 'Pending') {
```
When would be in a situation where the status is Pending? Doesn't the handler wait for it to finish before returning? We probably need to move rec report generation to be an async operation, but not for this PR. Something to keep in mind.

jennyhliu replied:
I updated, the lambda should wait.  It's waiting for the s3 file get updated.

jennyhliu replied:
Actually I don't understand the original logic, s3 reading should be eventually consistent. I removed it.

laurenfrederick commented on packages/api/tests/lambdas/test-db-indexer.js:
```diff
@@ -394,3 +412,47 @@ test.serial('The db-indexer does not throw an exception when execution fails', a
   );
 });
+
+test.serial('create, update and delete a reconciliationReport in DynamoDB and ES', async (t) => {
```
This test should be for the model and can be split up into a few tests. The test in this file should be a super simple test of just the dbIndexer functionality.

jennyhliu replied:
Removed the test from dbIndexer.  It's not necessary to add to the model test, all basic functions from model base.

jennyhliu commented on packages/api/lambdas/create-reconciliation-report.js:
```diff
@@ -493,12 +503,13 @@ async function createReconciliationReport(params) {
-    .then(() => null);
+  }).promise();
+
+  return reconciliationReportModel.updateStatus({ name: reportRecord.name }, 'Generated');
```
Updated lambda to set status to `Failed` when report creation fails.

laurenfrederick commented on packages/api/lambdas/create-reconciliation-report.js:
```diff
@@ -414,22 +415,28 @@ async function reconciliationReportForCumulusCMR(bucketsConfig) {
+ * @param {moment} reportStartTime - the report start time
+ * @returns {string} s3 report key
+ */
+const s3ReportKey = (stackName, reportStartTime) => `${stackName}/reconciliation-reports/inventoryReport-${reportStartTime.toISOString()}.json`;
```
Can you please put a test in for this?

laurenfrederick commented on packages/api/lambdas/create-reconciliation-report.js:
```diff
@@ -497,20 +504,59 @@ async function createReconciliationReport(params) {
+  const { systemBucket, stackName } = params;
+  const reportStartTime = moment.utc();
+  const reportRecordName = `inventoryReport-${reportStartTime.format('YYYYMMDDTHHmmssSSS')}`;
+  const reportKey = s3ReportKey(stackName, reportStartTime);
```
Should we pass `reportRecordName` into `s3ReportKey` instead of the `reportStartTime`?

laurenfrederick commented on packages/api/lambdas/create-reconciliation-report.js:
```diff
@@ -497,20 +504,59 @@ async function createReconciliationReport(params) {
+    await createReconciliationReport({ ...params, reportStartTime });
+    await reconciliationReportModel.updateStatus({ name: reportRecord.name }, 'Generated');
+  } catch (e) {
+    log.debug(`createReconciliationReport error ${e}`);
```
```suggestion
    log.error(`Error creating reconciliation report ${reportRecordName}`, e);
```
